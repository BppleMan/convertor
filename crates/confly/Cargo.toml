[package]
name = "confly"
description.workspace = true
authors.workspace = true
version.workspace = true
edition.workspace = true
license.workspace = true
repository.workspace = true

[dependencies]
# 错误 / 诊断
color-eyre = { workspace = true }

# 本地 / 工作区 包
convertor = { workspace = true }

# CLI / 工具
clap = { workspace = true, features = ["default", "cargo", "derive"] }

# HTTP / 头部 处理
headers = { workspace = true }

# 日志 / 追踪
tracing = { workspace = true, features = ["attributes"] }

# URL / 编码 / 解析
url = { workspace = true, features = ["serde"] }

# 异步运行时 / Tokio
tokio = { workspace = true, features = ["fs", "rt-multi-thread", "signal", "macros"] }

[dev-dependencies]
# 本地测试工具
convertor = { workspace = true, features = ["testkit"] }

# 断言 / 快照测试
insta = { workspace = true, features = ["yaml"] }

[features]
update = []
