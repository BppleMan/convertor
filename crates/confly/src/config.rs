use convertor::config::Config;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

mod client_config;

pub use client_config::*;
use convertor::config::proxy_client::ProxyClient;

#[derive(Debug, Clone)]
#[derive(Serialize, Deserialize)]
pub struct ConflyConfig {
    #[serde(flatten)]
    pub common: Config,

    #[serde(flatten)]
    pub clients: HashMap<ProxyClient, ClientConfig>,
}
