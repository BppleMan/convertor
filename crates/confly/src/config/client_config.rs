use std::path::PathBuf;

use convertor::config::proxy_client::ProxyClient;
use regex::{Captures, Regex};
use serde::{Deserialize, Serialize};
use std::sync::LazyLock;

static ENV_VAR_REGEX: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"\$\{([A-Za-z0-9_]+)}"#).expect("Failed to compile environment variable regex"));

#[derive(Default, Debug, Clone, Eq, PartialEq, Hash)]
#[derive(Serialize, Deserialize)]
pub struct ClientConfig {
    config_dir: String,
    main_profile_name: String,
    raw_profile_name: Option<String>,
    raw_sub_name: Option<String>,
    rules_name: Option<String>,
    sub_logs_name: Option<String>,
}

impl ClientConfig {
    pub fn surge_template() -> Self {
        Self {
            config_dir: "${ICLOUD}/../iCloud~com~nssurge~inc/Documents/surge".to_string(),
            main_profile_name: "surge.conf".to_string(),
            raw_profile_name: Some("raw.conf".to_string()),
            raw_sub_name: Some("BosLife.conf".to_string()),
            rules_name: Some("rules.dconf".to_string()),
            sub_logs_name: Some("subscription_logs.js".to_string()),
        }
    }

    pub fn clash_template() -> Self {
        Self {
            config_dir: "${HOME}/.config/mihomo".to_string(),
            main_profile_name: "config.yaml".to_string(),
            ..Default::default()
        }
    }
}

impl ClientConfig {
    pub fn set_surge_dir(&mut self, surge_dir: String) {
        self.config_dir = surge_dir;
    }

    pub fn surge_dir(&self) -> PathBuf {
        expand_env_vars(&self.config_dir).into()
    }

    pub fn main_profile_path(&self) -> PathBuf {
        self.surge_dir().join(expand_env_vars(&self.main_profile_name))
    }

    pub fn raw_profile_path(&self) -> Option<PathBuf> {
        self.raw_profile_name
            .as_ref()
            .map(|name| self.surge_dir().join(expand_env_vars(name)))
    }

    pub fn raw_sub_path(&self) -> Option<PathBuf> {
        self.raw_sub_name
            .as_ref()
            .map(|name| self.surge_dir().join(expand_env_vars(name)))
    }

    pub fn rules_path(&self) -> Option<PathBuf> {
        self.rules_name
            .as_ref()
            .map(|name| self.surge_dir().join(expand_env_vars(name)))
    }

    pub fn sub_logs_path(&self) -> Option<PathBuf> {
        self.sub_logs_name
            .as_ref()
            .map(|name| self.surge_dir().join(expand_env_vars(name)))
    }
}

fn expand_env_vars(value: impl AsRef<str>) -> String {
    let value = value.as_ref();
    ENV_VAR_REGEX
        .replace_all(value, |caps: &Captures| {
            let name = &caps[1];
            match std::env::var(name) {
                Ok(value) => value,
                Err(_) => name.to_string(),
            }
        })
        .to_string()
}
