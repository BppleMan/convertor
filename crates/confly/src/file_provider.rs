use color_eyre::Result;
use color_eyre::eyre::{OptionExt, eyre};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::{Arc, RwLock};

#[derive(Clone)]
pub enum FileProvider {
    FileSystem,
    Memory(Arc<RwLock<HashMap<PathBuf, String>>>),
}

impl FileProvider {
    pub fn read(&self, path: impl AsRef<Path>) -> Result<String> {
        let content = match self {
            FileProvider::FileSystem => std::fs::read_to_string(path)?,
            FileProvider::Memory(dir) => dir
                .read()
                .map_err(|e| eyre!("{e}"))?
                .get(&path.as_ref().to_path_buf())
                .cloned()
                .ok_or_eyre(format!("File not found in memory: {}", path.as_ref().display()))?,
        };
        Ok(content)
    }

    pub fn write(&self, path: impl AsRef<Path>, content: impl Into<String>) -> Result<()> {
        match self {
            FileProvider::FileSystem => std::fs::write(path, content.into())?,
            FileProvider::Memory(mem) => {
                mem.write()
                    .map_err(|e| eyre!("{e}"))?
                    .insert(path.as_ref().to_path_buf(), content.into());
            }
        };
        Ok(())
    }
}
