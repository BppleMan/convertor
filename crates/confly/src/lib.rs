pub mod command;
pub mod config;
pub mod file_provider;
pub mod profile_patcher;
