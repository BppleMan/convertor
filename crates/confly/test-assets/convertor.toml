secret = "bppleman"
server = "http://127.0.0.1:8080/"

[subscription]
sub_url = "http://127.0.0.1:8080/subscription"
interval = 86400
strict = true

[subscription.headers]
User-Agent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML), like Gecko) Chrome/"
sec-ch-ua = '"Not)A;Brand";v="8", "Chromium";v="138", "Google Chrome";v="138"'

[surge]
config_dir = "surge"
main_profile = "surge.conf"
raw = "raw.conf"
raw_profile = "raw_profile.conf"
rules = "rules.dconf"

[clash]
config_dir = "clash"
main_profile = "config.yaml"
