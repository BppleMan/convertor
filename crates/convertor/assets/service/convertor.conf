server {
    listen 443 ssl;
    http2 on;
    server_name convertor.bppleman.cn;

    # SSL 证书
    ssl_certificate     /etc/ssl/certs/bppleman.cn.pem;
    ssl_certificate_key /etc/ssl/private/bppleman.cn.key;

    # SSL 加密推荐配置
    ssl_protocols TLSv1.2 TLSv1.3;
    ssl_prefer_server_ciphers on;
    ssl_ciphers HIGH:!aNULL:!MD5;

    # 专属日志路径
    access_log /var/log/nginx/convertor_access.log main;
    error_log  /var/log/nginx/convertor_error.log warn;

    # 反代设置
    location / {
        proxy_pass http://127.0.0.1:8080;

        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_buffering off;

        proxy_connect_timeout 10s;
        proxy_send_timeout    30s;
        proxy_read_timeout    30s;
    }

    # gzip 压缩
    gzip on;
    gzip_types text/plain application/json application/javascript text/css;

    # 安全 header（可按需删改）
    add_header X-Content-Type-Options nosniff;
    add_header X-Frame-Options DENY;
    add_header X-XSS-Protection "1; mode=block";
}
