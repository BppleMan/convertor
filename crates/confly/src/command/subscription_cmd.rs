use crate::config::ConflyConfig;
use crate::file_provider::FileProvider;
use clap::Args;
use color_eyre::Result;
use color_eyre::eyre::OptionExt;
use convertor::common::encrypt::encrypt;
use convertor::config::proxy_client::ProxyClient;
use convertor::core::profile::Profile;
use convertor::core::profile::clash_profile::ClashProfile;
use convertor::core::profile::extract_policies_for_rule_provider;
use convertor::core::profile::policy::Policy;
use convertor::core::profile::surge_profile::SurgeProfile;
use convertor::error::UrlBuilderError;
use convertor::provider::SubsProvider;
use convertor::url::url_builder::{HostPort, UrlBuilder};
use convertor::url::url_result::UrlResult;
use url::Url;

#[derive(Default, Debug, Clone, Hash, Args)]
pub struct SubscriptionCmd {
    /// 构造适用于不同客户端的订阅地址
    #[arg(value_enum)]
    pub client: ProxyClient,

    /// 原始订阅链接(raw_url)
    #[arg()]
    pub url: Option<Url>,

    /// 是否更新本地订阅文件
    #[arg(short, long, default_value_t = false)]
    pub update: bool,
}

#[allow(clippy::large_enum_variant)]
enum ClientProfile {
    Surge,
    Clash(ClashProfile),
}

impl SubscriptionCmd {
    pub async fn execute(
        self,
        config: &ConflyConfig,
        subs_provider: &SubsProvider,
        file_provider: &FileProvider,
    ) -> Result<(UrlBuilder, UrlResult)> {
        let url_builder = self.create_url_builder(config).await?;
        let raw_url = url_builder.build_raw_url();
        let raw_profile_content = subs_provider
            .get_raw_profile(raw_url.into(), [("User-Agent", "Surge Mac/8310")].into())
            .await?;
        let sub_host = url_builder
            .sub_url
            .host_port()
            .ok_or_eyre("无法从 sub_url 中提取 host port")?;
        let (client_profile, policies) = match self.client {
            ProxyClient::Surge => {
                let mut raw_profile = SurgeProfile::parse(raw_profile_content)?;
                raw_profile.convert(&url_builder)?;
                let mut policies: Vec<Policy> = raw_profile.policy_of_rules.keys().cloned().collect();
                policies.sort();
                (ClientProfile::Surge, policies)
            }
            ProxyClient::Clash => {
                let raw_profile = ClashProfile::parse(raw_profile_content)?;
                let policies = extract_policies_for_rule_provider(&raw_profile.rules, sub_host);
                (ClientProfile::Clash(raw_profile), policies)
            }
        };

        let raw_url = url_builder.build_raw_url();
        let profile_url = url_builder.build_profile_url()?;
        let raw_profile_url = url_builder.build_raw_profile_url()?;
        let rule_provider_urls = policies
            .iter()
            .map(|policy| url_builder.build_rule_provider_url(policy))
            .collect::<Result<Vec<_>, UrlBuilderError>>()?;
        let result = UrlResult {
            raw_url,
            raw_profile_url,
            profile_url,
            rule_providers_url: rule_provider_urls,
        };

        // 副作用逻辑后置，主流程只负责数据流
        if self.update {
            match (client_profile, config.clients.get(&self.client)) {
                (ClientProfile::Surge, Some(client_config)) => {
                    client_config.update_surge_config(file_provider, &url_builder, &policies)?;
                }
                (ClientProfile::Clash(profile), Some(client_config)) => {
                    client_config.update_clash_config(file_provider, &url_builder, profile, &config.common.secret)?;
                }
                _ => eprintln!("未找到对应的客户端配置，跳过更新本地订阅文件"),
            }
        }
        Ok((url_builder, result))
    }

    async fn create_url_builder(&self, config: &ConflyConfig) -> Result<UrlBuilder> {
        let SubscriptionCmd { client, url, update: _ } = &self;
        let subscription_config = &config.common.subscription;

        let sub_url = url.clone().unwrap_or(subscription_config.sub_url.clone());
        let server = config.common.server.clone();
        let secret = config.common.secret.clone();
        let enc_secret = encrypt(secret.as_bytes(), &secret)?;
        let enc_sub_url = encrypt(secret.as_bytes(), sub_url.as_str())?;
        let interval = subscription_config.interval;
        let strict = subscription_config.strict;

        let url_builder = UrlBuilder::new(
            secret,
            Some(enc_secret),
            *client,
            server,
            sub_url,
            Some(enc_sub_url),
            interval,
            strict,
        )?;

        Ok(url_builder)
    }
}

// impl SubscriptionCmd {
//     pub fn snapshot_name(&self) -> String {
//         let client = self.client.to_string();
//         let url = self
//             .url
//             .as_ref()
//             .map_or("no_url".to_string(), |url| url.host_port().unwrap());
//         let server = self
//             .c
//             .as_ref()
//             .map_or("no_server".to_string(), |server| server.to_string());
//         let interval = self
//             .interval
//             .map_or("no_interval".to_string(), |interval| interval.to_string());
//         let strict = self.strict.map_or("no_strict".to_string(), |_| "strict".to_string());
//         let reset = if self.reset { "reset" } else { "no_reset" };
//
//         let update = if self.update { "update" } else { "no_update" };
//
//         format!("{client}-{provider}-{url}-{server}-{interval}-{strict}-{reset}-{update}")
//     }
// }
