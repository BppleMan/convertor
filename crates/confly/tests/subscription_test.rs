use color_eyre::eyre::OptionExt;
use confly::command::subscription_cmd::SubscriptionCmd;
use confly::config::{ClientConfig, ConflyConfig};
use confly::file_provider::FileProvider;
use convertor::config::proxy_client::ProxyClient;
use convertor::init_test;
use convertor::provider::SubsProvider;
use convertor::testkit::start_mock_provider_server;
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

pub fn cmds(client: ProxyClient) -> [SubscriptionCmd; 2] {
    [
        SubscriptionCmd {
            client,
            url: None,
            update: false,
        },
        SubscriptionCmd {
            client,
            url: None,
            update: true,
        },
    ]
}

pub fn file_provider(config: &ClientConfig) -> FileProvider {
    let mut test_assets_dir = HashMap::new();
    test_assets_dir.insert(config.main_profile_path(), "".to_string());
    if let Some(raw_path) = config.raw_path() {
        test_assets_dir.insert(raw_path, "".to_string());
    }
    if let Some(raw_profile_path) = config.raw_profile_path() {
        test_assets_dir.insert(raw_profile_path, "".to_string());
    }
    if let Some(rules_path) = config.rules_path() {
        test_assets_dir.insert(
            rules_path,
            "# Rule Provider from convertor\n# End of Rule Provider".to_string(),
        );
    }
    FileProvider::Memory(Arc::new(RwLock::new(test_assets_dir)))
}

async fn test_subscription(client: ProxyClient) -> color_eyre::Result<()> {
    let base_dir = init_test!();
    let mut config = ConflyConfig::search(&base_dir, None::<&str>)?;
    let client_config = config
        .clients
        .get(&client)
        .ok_or_eyre(format!("没有找到 {client} 客户端配置"))?;
    start_mock_provider_server(&mut config.common).await?;

    let subs_provider = SubsProvider::new(None, config.common.redis.as_ref().map(|r| r.prefix.as_str()));
    let cmds = cmds(client);
    for (i, cmd) in cmds.into_iter().enumerate() {
        let ctx = format!("test_subscription_{client}_cmd_{i}");
        let file_provider = file_provider(client_config);
        let (url_builder, result) = cmd.clone().execute(&config, &subs_provider, &file_provider).await?;
        let result = result.to_string();
        let result = result
            .replace(
                &url_builder
                    .sub_url
                    .port()
                    .map(|p| p.to_string())
                    .unwrap_or("".to_string()),
                "<PORT>",
            )
            .replace(&url_builder.server.to_string(), "<SERVER>")
            .replace(&url_builder.enc_sub_url, "<ENC_SUB_URL>");
        insta::assert_snapshot!(ctx, result);
        if cmd.update {
            insta::assert_snapshot!(
                client_config.main_profile_path().display().to_string(),
                file_provider
                    .read(client_config.main_profile_path())?
                    .replace(&url_builder.server.to_string(), "<SERVER>")
                    .replace(&url_builder.enc_sub_url, "<ENC_SUB_URL>")
            );
            if let Some(raw_path) = client_config.raw_path() {
                insta::assert_snapshot!(
                    raw_path.display().to_string(),
                    file_provider
                        .read(&raw_path)?
                        .replace(&url_builder.server.to_string(), "<SERVER>")
                        .replace(url_builder.sub_url.as_str(), "<SUB_URL>")
                        .replace(&url_builder.enc_sub_url, "<ENC_SUB_URL>")
                );
            }
            if let Some(raw_profile_path) = client_config.raw_profile_path() {
                insta::assert_snapshot!(
                    raw_profile_path.display().to_string(),
                    file_provider
                        .read(&raw_profile_path)?
                        .replace(&url_builder.server.to_string(), "<SERVER>")
                        .replace(&url_builder.enc_sub_url, "<ENC_SUB_URL>")
                );
            }
            if let Some(rules_path) = client_config.rules_path() {
                insta::assert_snapshot!(
                    rules_path.display().to_string(),
                    file_provider
                        .read(&rules_path)?
                        .replace(&url_builder.server.to_string(), "<SERVER>")
                        .replace(&url_builder.enc_sub_url, "<ENC_SUB_URL>")
                );
            }
        }
    }
    Ok(())
}

#[tokio::test]
async fn test_subscription_surge() -> color_eyre::Result<()> {
    test_subscription(ProxyClient::Surge).await
}

#[tokio::test]
async fn test_subscription_clash() -> color_eyre::Result<()> {
    test_subscription(ProxyClient::Clash).await
}
