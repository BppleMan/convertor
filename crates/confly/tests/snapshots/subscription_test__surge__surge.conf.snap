---
source: crates/confly/tests/subscription_test.rs
expression: "file_provider.read(client_config.main_profile_path())?\n.replace(&url_builder.server.to_string(),\n\"<SERVER>\").replace(&url_builder.enc_sub_url, \"<ENC_SUB_URL>\")"
---
#!MANAGED-CONFIG <SERVER>profile/surge?interval=86400&strict=true&sub_url=<ENC_SUB_URL> interval=86400 strict=true
